File: packages/db/src/query/sort.ts
import { SchemaTypeDef } from '../schema/types.js'
import { Query } from './query.js'

export const createSortBuffer = (
  schema: SchemaTypeDef,
  field: string,
  order?: 'asc' | 'desc',
) => {
  const fieldDef = schema.props[field]
  if (!fieldDef) {
    console.warn('Query: No field def defined for', field)
    return this
  }

  if (order !== undefined && fieldDef.prop === 0) {
    const buf = Buffer.allocUnsafe(7)
    buf[0] = order === 'asc' ? 0 : 1
    buf[1] = 0
    buf[2] = fieldDef.typeIndex
    buf.writeUint16LE(fieldDef.start, 3)
    buf.writeUint16LE(fieldDef.len, 5)
    return buf
  }

  if (order !== undefined) {
    const buf = Buffer.allocUnsafe(3)
    buf[0] = order === 'asc' ? 0 : 1
    buf[1] = fieldDef.prop
    buf[2] = fieldDef.typeIndex
    return buf
  }

  if (fieldDef.prop === 0) {
    const buf = Buffer.allocUnsafe(6)
    buf[0] = 0
    buf[1] = fieldDef.typeIndex
    buf.writeUint16LE(fieldDef.start, 2)
    buf.writeUint16LE(fieldDef.len, 4)
    return buf
  }

  const buf = Buffer.allocUnsafe(2)
  buf[0] = fieldDef.prop
  buf[1] = fieldDef.typeIndex
  return buf
}

export const sort = (
  query: Query,
  field: string,
  order: 'asc' | 'desc' = 'asc',
) => {
  query.sortOrder = order === 'asc' ? 0 : 1
  query.sortBuffer = createSortBuffer(query.schema, field)
}

File: packages/db/native/query/types/sort.zig
const c = @import("../../c.zig");
const errors = @import("../../errors.zig");
const db = @import("../../db/db.zig");
const selva = @import("../../selva.zig");
const getFields = @import("../include/include.zig").getFields;
const results = @import("../results.zig");
const QueryCtx = @import("../types.zig").QueryCtx;
const filter = @import("../filter/filter.zig").filter;
const sort = @import("../../db/sort.zig");
const types = @import("../../types.zig");
const hasId = @import("../hasId.zig").hasId;
const searchStr = @import("../filter/search.zig");
const readInt = @import("../../utils.zig").readInt;
const Result = @import("../results.zig").Result;
const s = @import("./search.zig");
const std = @import("std");

pub fn default(
    comptime desc: bool,
    ctx: *QueryCtx,
    offset: u32,
    limit: u32,
    typeId: db.TypeId,
    conditions: []u8,
    include: []u8,
    sortBuffer: []u8,
) !void {
    // [order] [prop] [propType] [start] [start] [len] [len]
    const field = sortBuffer[0];
    const start = readInt(u16, sortBuffer, 2);
    const sIndex = sort.getSortIndex(ctx.db.sortIndexes.get(typeId), field, start);
    if (sIndex == null) {
        std.log.err(
            "Err exec query (zig) no sort index aviable for query type: {any} field: {any} start: {any}  \n",
            .{ typeId, field, start },
        );
        return;
    }
    const typeEntry = try db.getType(ctx.db, typeId);
    const sI = sIndex.?;
    if (desc) {
        selva.selva_sort_foreach_begin_reverse(sI.index);
    } else {
        selva.selva_sort_foreach_begin(sI.index);
    }
    var correctedForOffset: u32 = offset;
    checkItem: while (!selva.selva_sort_foreach_done(sI.index)) {
        var node: db.Node = undefined;
        if (desc) {
            node = @ptrCast(selva.selva_sort_foreach_reverse(sI.index));
        } else {
            node = @ptrCast(selva.selva_sort_foreach(sI.index));
        }
        if (!filter(ctx.db, node, typeEntry, conditions, null, null, 0, false)) {
            continue :checkItem;
        }
        if (correctedForOffset != 0) {
            correctedForOffset -= 1;
            continue :checkItem;
        }
        const id = db.getNodeId(node);
        const size = try getFields(node, ctx, id, typeEntry, include, null, null, false);
        if (size > 0) {
            ctx.size += size;
            ctx.totalResults += 1;
        }
        if (ctx.totalResults >= limit) {
            break;
        }
    }
}

pub fn search(
    comptime desc: bool,
    ctx: *QueryCtx,
    offset: u32,
    limit: u32,
    typeId: db.TypeId,
    conditions: []u8,
    include: []u8,
    sortBuffer: []u8,
    searchCtx: *const searchStr.SearchCtx,
) !void {
    // [order] [prop] [propType] [start] [start] [len] [len]
    const field = sortBuffer[0];
    const start = readInt(u16, sortBuffer, 2);
    const sIndex = sort.getSortIndex(ctx.db.sortIndexes.get(typeId), field, start);
    if (sIndex == null) {
        std.log.err(
            "Err exec query (zig) no sort index aviable for query type: {any} field: {any} start: {any}  \n",
            .{ typeId, field, start },
        );
        return;
    }
    const typeEntry = try db.getType(ctx.db, typeId);
    const sI = sIndex.?;
    if (desc) {
        selva.selva_sort_foreach_begin_reverse(sI.index);
    } else {
        selva.selva_sort_foreach_begin(sI.index);
    }
    var searchCtxC = s.createSearchCtx(offset);
    while (!selva.selva_sort_foreach_done(sI.index)) {
        var node: db.Node = undefined;
        if (desc) {
            node = @ptrCast(selva.selva_sort_foreach_reverse(sI.index));
        } else {
            node = @ptrCast(selva.selva_sort_foreach(sI.index));
        }
        s.addToScore(ctx, &searchCtxC, node, typeEntry, conditions, searchCtx);
        if ((searchCtxC.totalSearchResults >= limit)) {
            break;
        }
    }
    try s.addToResults(ctx, &searchCtxC, include, limit, typeEntry);
}
